scalalib/worker/src/mill/scalalib/worker/ZincWorkerImpl.scala

```diff
@@ -708,4 +684,38 @@ object ZincWorkerImpl {
       toAnalyze = List((List(start), deps(start).toList))
     ).reverse
   }
+
+  /**
+   * In memory cache backed by a ConcurrentHashMap[A, SoftReference[B]]
+   * `SoftReference`s allow the GC to free values when memory is scarce
+   */
+  private class Cache[A, B <: AnyRef] {
+    protected val cache =
+      new java.util.concurrent.ConcurrentHashMap[A, SoftReference[B]]
+
+    def getOrElseCreate(a: A)(create: => B): B =
+      cache.compute(
+        a,
+        {
+          case (_, v @ SoftReference(_)) => v
+          case _ => SoftReference(create)
+        }
+      )()
+
+    import scala.jdk.CollectionConverters._
+    def clear(closeAutoCloseables: Boolean): Unit = {
+
+      if (closeAutoCloseables) {
+        cache
+          .values()
+          .iterator()
+          .asScala
+          .foreach { case SoftReference(v: AutoCloseable) =>
```

Comment:
`foreach` doesn't take a partial function, so we should provide a catch-all `case`. Not doing so will fail for non-`Autoclosable` entries.